### PR TITLE
Add the mimalloc option to speed up the compiler on Windows

### PR DESCRIPTION
**Pull request by hjyamauchi**

### CODE REVIEW

File: .github/workflows/swift-toolchain.yml [hunk 1]
```diff
@@ -3193,6 +3193,34 @@ jobs:
             Install-Package -Name WixToolset.Sdk -RequiredVersion 4.0.1 -Force
           }
 
+      - uses: actions/checkout@v4
+        if: matrix.arch == 'amd64'
+        with:
+          repository: microsoft/mimalloc
+          ref: refs/tags/v2.1.7
```
Comment (compnerd) [inline .github/workflows/swift-toolchain.yml:3200]: Can we pull this out into a global like we have done for the other dependencies (thanks to @Steelskin!).

File: .github/workflows/swift-toolchain.yml [hunk 2]
```diff
@@ -3193,6 +3193,34 @@ jobs:
...
+          path: ${{ github.workspace }}/SourceCache/mimalloc
+          show-progress: false
+
+      - name: Build and apply mimalloc
+        if: matrix.arch == 'amd64'
```
Comment (compnerd) [inline .github/workflows/swift-toolchain.yml:3205]: Should we be checking the OS as well as we are working towards the macOS toolchain?

File: .github/workflows/swift-toolchain.yml [hunk 3]
```diff
@@ -3193,6 +3193,34 @@ jobs:
...
+          # Reference: https://github.com/microsoft/mimalloc/tree/dev/bin#minject
+          msbuild ${{ github.workspace }}\SourceCache\mimalloc\ide\vs2022\mimalloc.sln -p:Configuration=Release
+          $ToolchainBin = "${{ github.workspace }}\BuildRoot\Library\Developer\Toolchains\unknown-Asserts-development.xctoolchain\usr\bin"
+          Copy-Item -Path "${{ github.workspace }}\SourceCache\mimalloc\out\msvc-x64\Release\mimalloc-override.dll" `
+            -Destination "$ToolchainBin"
+          Copy-Item -Path "${{ github.workspace }}\SourceCache\mimalloc\out\msvc-x64\Release\mimalloc-redirect.dll" `
+            -Destination "$ToolchainBin"
+          $MimallocExecutables = @("swift.exe","swiftc.exe","swift-driver.exe","swift-frontend.exe")
+          $MimallocExecutables += @("clang.exe","clang++.exe","clang-cl.exe")
+          $MimallocExecutables += @("lld.exe","lld-link.exe","ld.lld.exe","ld64.lld.exe")
```
Comment (compnerd) [inline .github/workflows/swift-toolchain.yml:3216]: Why not create a single list with a single element per line?

File: default.xml [hunk 4]
```diff
@@ -64,4 +64,6 @@
   <project remote="github" name="madler/zlib" path="zlib" revision="refs/tags/v1.3.1" />
 
   <project remote="github" name="ninja-build/ninja" path="ninja" groups="notdefault,dependencies" revision="master" />
+
+  <project remote="github" name="microsoft/mimalloc" path="mimalloc" revision="refs/tags/v2.1.7" />
```
Comment (compnerd) [inline default.xml:68]: The other vendors are alphabetical, can we hoist this above `ninja-build`?

File: .github/workflows/swift-toolchain.yml (hunk 1 above)
Comment (hjyamauchi) [reply .github/workflows/swift-toolchain.yml]: Done

File: .github/workflows/swift-toolchain.yml (hunk 2 above)
Comment (hjyamauchi) [reply .github/workflows/swift-toolchain.yml:3208]: As the entire enclosing job is restricted to Windows (and the matrix doesn't have OSes yet), so this looks okay to me for now.
```
  package_tools:
    # TODO: Build this on macOS or make an equivalent Mac-only job
    if: inputs.build_os == 'Windows'
    ...
```

File: .github/workflows/swift-toolchain.yml (hunk 3 above)
Comment (hjyamauchi) [reply .github/workflows/swift-toolchain.yml]: Done

File: default.xml (hunk 4 above)
Comment (hjyamauchi) [reply default.xml:66]: Done

**Review by compnerd — APPROVED** [review]

**Pull request merged**